File: compose/shared/src/commonMain/kotlin/me/batashev/stride/data/Session.kt
package me.batashev.stride.data

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import me.batashev.stride.Settings

/**
 * Holds the cloud server base URL and bearer token for the signed-in user and
 * mirrors them to persistent [Settings]. The base URL is kept after sign-out so
 * the login screen can prefill it.
 */
class Session(private val settings: Settings) {

    data class Auth(val baseUrl: String?, val token: String?) {
        val isAuthenticated: Boolean
            get() = !baseUrl.isNullOrBlank() && !token.isNullOrBlank()
    }

    private val _state = MutableStateFlow(read())
    val state: StateFlow<Auth> = _state.asStateFlow()

    val baseUrl: String? get() = _state.value.baseUrl
    val token: String? get() = _state.value.token
    val isAuthenticated: Boolean get() = _state.value.isAuthenticated

    fun signIn(baseUrl: String, token: String) {
        settings.putString(KEY_BASE, baseUrl)
        settings.putString(KEY_TOKEN, token)
        _state.value = Auth(baseUrl, token)
    }

    fun signOut() {
        settings.remove(KEY_TOKEN)
        _state.value = _state.value.copy(token = null)
    }

    private fun read() = Auth(settings.getString(KEY_BASE), settings.getString(KEY_TOKEN))

    private companion object {
        const val KEY_BASE = "friday.baseURL"
        const val KEY_TOKEN = "friday.token"
    }
}

File: apple/Stride/Stride/Friday/Core/Session.swift
import Foundation

/// Holds the cloud server location and bearer token for the signed-in user and
/// mirrors them to `UserDefaults` so a session survives relaunches. The base URL
/// is kept after sign-out so the login screen can prefill it.
final class Session: @unchecked Sendable {
    static let shared = Session()

    private let lock = NSLock()
    private var storedBaseURL: URL?
    private var storedToken: String?

    private enum Key {
        static let baseURL = "friday.baseURL"
        static let token = "friday.token"
    }

    private init() {
        let defaults = UserDefaults.standard
        if let raw = defaults.string(forKey: Key.baseURL) {
            storedBaseURL = URL(string: raw)
        }
        storedToken = defaults.string(forKey: Key.token)
    }

    var baseURL: URL? {
        lock.withLock { storedBaseURL }
    }

    var token: String? {
        lock.withLock { storedToken }
    }

    var isAuthenticated: Bool {
        lock.withLock { storedBaseURL != nil && storedToken != nil }
    }

    func signIn(baseURL: URL, token: String) {
        lock.withLock {
            storedBaseURL = baseURL
            storedToken = token
        }
        let defaults = UserDefaults.standard
        defaults.set(baseURL.absoluteString, forKey: Key.baseURL)
        defaults.set(token, forKey: Key.token)
    }

    func signOut() {
        lock.withLock { storedToken = nil }
        UserDefaults.standard.removeObject(forKey: Key.token)
    }
}
